state-chain/pallets/cf-asset-balances/src/lib.rs

```diff
@@ -0,0 +1,312 @@
+		amount_owed.defensive_saturating_reduce(amount_reconciled);
+
+		Ok(())
+	}
+
+	pub fn trigger_reconciliation() {
+		if !T::SafeMode::get().reconciliation_enabled {
+			log::info!("Reconciliation is disabled. Skipping reconciliation.");
+			return;
+		}
+
+		for chain in ForeignChain::iter() {
+			let owed_assets = Liabilities::<T>::take(chain.gas_asset());
+			WithheldAssets::<T>::mutate(chain.gas_asset(), |total_withheld| {
+				let mut owed_assets = owed_assets.into_iter().collect::<Vec<_>>();
+				owed_assets.sort_by_key(|(_, amount)| core::cmp::Reverse(*amount));
+
+				for (destination, amount) in &mut owed_assets {
+					debug_assert!(*amount > 0);
+					match Self::reconcile(chain, destination, amount, total_withheld) {
+						Ok(_) =>
+							if *total_withheld == 0 {
+								break;
+							},
+						Err(_) => break,
+					}
+				}
+
+				owed_assets.retain(|(_, amount)| *amount > 0);
+				if !owed_assets.is_empty() {
+					Liabilities::<T>::insert(
+						chain.gas_asset(),
+						owed_assets.into_iter().collect::<BTreeMap<_, _>>(),
+					);
+				}
+			});
+		}
+	}
+
+	pub fn vault_imbalance(asset: Asset) -> VaultImbalance<AssetAmount> {
+		let owed = Liabilities::<T>::get(asset).values().sum::<u128>();
+		let withheld = WithheldAssets::<T>::get(asset);
+		if owed > withheld {
+			VaultImbalance::Deficit(owed - withheld)
+		} else {
+			VaultImbalance::Surplus(withheld - owed)
+		}
+	}
+}
+
+#[derive(Encode, Decode, TypeInfo, Clone, PartialEq, Eq, RuntimeDebug, Serialize, Deserialize)]
+pub enum VaultImbalance<A> {
+	/// There are more withheld assets than what is owed.
+	Surplus(A),
+	/// There are more assets owed than what is withheld.
+	Deficit(A),
+}
+
+impl<A> VaultImbalance<A> {
+	pub fn map<B>(&self, f: impl FnOnce(&A) -> B) -> VaultImbalance<B> {
+		match self {
+			VaultImbalance::Surplus(amount) => VaultImbalance::Surplus(f(amount)),
+			VaultImbalance::Deficit(amount) => VaultImbalance::Deficit(f(amount)),
+		}
+	}
+}
+
+impl<T: Config> LiabilityTracker for Pallet<T> {
+	fn record_liability(address: ForeignChainAddress, asset: Asset, amount: AssetAmount) {
+		debug_assert_eq!(ForeignChain::from(asset), address.chain());
+		Liabilities::<T>::mutate(asset, |fees| {
+			fees.entry(match ForeignChain::from(asset) {
+				ForeignChain::Ethereum | ForeignChain::Arbitrum => address.into(),
+				ForeignChain::Polkadot => ExternalOwner::AggKey,
+				ForeignChain::Bitcoin | ForeignChain::Solana => ExternalOwner::Vault,
+			})
+			.and_modify(|fee| fee.saturating_accrue(amount))
+			.or_insert(amount);
+		});
+	}
+
+	#[cfg(feature = "try-runtime")]
+	fn total_liabilities(asset: Asset) -> u128 {
+		Liabilities::<T>::get(asset).values().sum()
+	}
+}
+
+impl<T: Config> AssetWithholding for Pallet<T> {
+	fn withhold_assets(asset: Asset, amount: AssetAmount) {
+		WithheldAssets::<T>::mutate(asset, |fees| {
+			fees.saturating_accrue(amount);
+		});
+	}
+
+	#[cfg(feature = "try-runtime")]
```

Comment:
I think because we need to access the storage in a migration of another pallet for the verification (also the reason for the try-runtime feature). I think that's the easiest we to do it also in terms of removing the references to the other pallet after the migration is done.